# Basic types and operations

## $ \S 1 $ Objects, classes, values, variables and assignment

### $ 1.1 $ Basic terminology of object-oriented programming

A __class__ (or _data type_) consists of a set of values and a set of
permissible operations on these values. When a class is already built into
Python, it is usually called a __type__.  For example, in Python there are types
for the representation of integers, real numbers, text strings and booleans,
among others. 

For example, an integer can take on the value $ -3 $, but not the value $ 3.14
$. Booleans can only take on two values: `True` or `False`.  Similarly, we can
multiply two real numbers, but not one string by another; we can concatenate
two strings, but that doesn't make sense for boolean values.

The integer $ -3 $ is an _object_ belonging to the class of integers. More
generally, in programming, an __object__ is an aggregate abstraction of some
data. For example, we may choose to represent a car in Python by a set of values
such as its manufacturer (a string), model (also a string), its year of
manufacture (an integer), its color, its identification number, etc., all
encapsulated as a unit. Concretely, an object is just an instance of a specific
class residing in memory. In Python, everything is represented is an object,
and every object must belong to a class.

As another example, if we had to write software for a banking system, it would in
principle be possible to work only with built-in types. However, this would be
extremely cumbersome. In order to encapsulate the data and behavior associated
with the manipulation of bank accounts at a higher level of abstraction, we
should design and implement an appropriate class ourselves, using built-in types
and other previously existing classes and operations as building blocks. Each
bank account in our system would then be an object belonging to, or an instance
of, this class.

📝 In the terminology of object-oriented programming, the technical terms for
"property" and "operation" are _attribute_ and _method_, respectively. Thus,
the purpose of a class is to implement a specific set of attributes and
methods to be inherited by every object (instance) of that class. The collection
of values assumed by all the attributes of an object is called its _state_.

In the example of the bank account class, two of the attributes could be
the current balance and the account holder's ID. There would probably be
methods designed to model deposits, withdrawals, transfers, modifications in
the holder's address, etc. 

In this tutorial we will not need to discuss objects, classes and object-oriented
programming beyond what has already been mentioned. In fact, we will work only
with built-in types.

### $ 1. 2 $ Variables

Each object created by the Python interpreter is assigned a unique memory
address that we never need to handle directly or worry about. That is because
this address is abstracted away through the use of _variables_.  A __variable__
is an association of an __identifier__ (i.e., name or label) with an object.
In Python, variables are created through an __assignment__ statement, using the
__assignment operator__ `=` in the form _\<identifier>_ `=` _\<object>_.  This
binds the identifier to the left of the `=` sign to the object to its right.

In [1]:
a = 1729   # Bind the identifier 'a' to the integer 1729.
a          # Return the value of the corresponding object as output of the cell.

1729

📝 Unlike some other programming languages, in Python the syntax for _creating
a new variable_ and for _assigning a new object to an existing variable name_ is
exactly the same:

In [2]:
b = 1729       # Create a new variable b and assign the object 1729 to it.
print(b)       # Print the current value of b.
b = 17.29      # Reassign a _new_ object to the name 'b'.
print(b)       # Print the current value of b.

1729
17.29


Identifiers (that is, variable names) may consist only of letters,
digits and __underscores__ \_; the initial character must be a letter or
underscore. The class of "letters" actually includes many non-Latin Unicode
characters.


In [13]:
variable_1 = 10     # Variable name using letters, underscore and a digit
变量 = "Chinese"    # Chinese characters
π = 3.14            # Greek letter

print(variable_1)
print(变量)
print(π)

10
Chinese
3.14


In [9]:
# Unfortunately, we can't yet use emojis as variable names 🙁:
🚀 = "rocket"

SyntaxError: invalid character '🚀' (U+1F680) (4263400734.py, line 2)

📝 By convention, uppercase initial letters are usually reserved to denote
classes and lowercase initial letters to name objects. The identifiers
of constant values are frequently written using uppercase characters only.

<div class="alert alert-info"> In Python, simply creating a variable or assigning it a new object does not prompt the interpreter to return nor print its value.
    <ul>
        <li>To get the interpreter to <i>return</i> the value of the object
        bound to a variable, say $ x $, as the output of a cell, one can use a
        statement consisting solely of its name, in this case <code>x</code>.</li>
        <li>To <i>print</i> the value of the object referred to by a variable,
        say $ x $, to the screen, use <code>print(x)</code>. Note that calling
        <code>print</code> <i>generates no output</i> (more precisely, it
        returns <code>None</code>, as we will discuss later).</li></ul></div>

More generally, the function `print` can be used to print to the screen any collection of arguments, separated by commas. These commas will be replaced by single spaces when the arguments are displayed.

In [2]:
a = 'cunning'
b = 'wizard'
print(a)
print(a, b)

cunning
cunning wizard


⚠️ If a code cell contains more than one statement which produces an output,
then only the result of the most recent one will be yielded as the output of the
cell; the other ones are discarded. 

In [10]:
# This statement causes the interpreter to evaluate the variable a:
a    # (evaluates to the string 'cunning')
# However, if we issue another such statement:
b    # (evaluates to the string 'wizard')
# then only the latter will actually yield an output.

'wizard'

The type of a variable $ x $ can be inspected through a call to the function
`type`, as in `type(x)`. 

In [1]:
number = 'two'  # The string 'two' is assigned to the new variable `number`.
print(number, type(number))

number = 2  # The variable `number` now points to an object of different type.
print(number, type(number))

two <class 'str'>
2 <class 'int'>


As in the preceding example, the Python interpreter automatically infers the
type of a variable from its value; this feature is called __type inference__.
Moreover, in Python (in contrast to, say, C or Java), variables are
__typed dinamically__, meaning that not only can the _object_ bound to a
variable change during a program, but even its _class_ (or _type_) can be
modified. Whenever a new object is assigned to a previously used identifier, the
original association is lost. 

📝 Blank lines in a code cell are ignored by the interpreter. We can use them to
visually separate the code into several parts in order to improve legibility.

The statement `isinstance(<variable>, <class>)` returns either `True` or `False`
according to whether a given variable belongs to the indicated class or not.

__Exercise:__ If the statement `lifespan = 120` is run, determine the
values of the following (here `str`, `int` and `float` are the names of the
types of strings, integers and floating-point numbers, respectively):

(a) `isinstance(lifespan, str)`

(b) `isinstance(lifespan, int)`

(c) `isinstance(lifespan, float)`

How do the answers change if `lifespan = 120.0`?

In [2]:
lifespan = 120

We can also determine whether an object belongs to one of several types, as in the following example:

In [3]:
x = 12.4
isinstance(x, (str, float))  # Returns `True` since x is a float

True

⚠️ All names (not just those of variables) are __case-sensitive__. Thus `Staff` and `staff` may refer to completely different objects.

__Exercise:__ Predict what gets printed when you run the code cell below.

In [ ]:
Mammal = "whale"
mammal = "dog"

print(Mammal)

__Exercise:__ Can you explain what happens if the statement `print(print(p))` is interpreted?

In [4]:
p = "print me"

__Exercise:__ What happens if you issue the following statements consecutively?
What is the output (if any) in each case?

(a) `a = 2`

(b) `b = 3`

(c) `c = a * b`

(d) `print(c)`

(e) `a - b`

(f) `a**b`

🚫 The following small set of keywords cannot be used to name objects because they already have a special meaning to the interpreter (don't try to memorize it!):

|         |            |         |          |
|:-------:|:----------:|:-------:|:--------:|
| `False` | `None`     | `True`  | `and`    |
| `as`    | `assert`   | `async` | `await`  |
| `break` | `class`    | `continue` | `def` |
| `del`   | `elif`     | `else`  | `except` |
| `finally` | `for`    | `from`  | `global` |
| `if`    | `import`   | `in`    | `is`     |
| `lambda`| `nonlocal` | `not`   | `or`     |
| `pass`  | `raise`    | `return`| `try`    |
| `while` | `with`     | `yield` |          |


__Exercise:__

(a) Let `x` and `y` have values $ 1 $ and $ 2 $, respectively. Suppose that we
want to swap their values. Can we accomplish this through the instructions in the code
cell below? Explain.

In [3]:
x = 1    # Initializing x.
y = 2    # Initializing y.

# Performing the swap:
x = y
y = x
print(x, y)

(b) How could one use a third, temporary variable `temp` to solve this problem?

In Python, one can make several __simultaneous assignments__ in a single line
using commas `,` as separators. This is especially useful for permuting the
values of two or more variables without resorting to temporary variables.
However, as a best practice one should avoid needlessly using this feature
because it hampers the legibility of code.

__Example:__

In [45]:
x = 1
y = 2
print(x, y)

x, y = y, x
print(x, y)

1 2
2 1


In [6]:
# Multiple assignments are made _simultaneously_.
# Check this in the following example:
x = 1
y = 2
x, y = x + y, x - y

__Exercise:__ As we have said before, in Python every object must belong to
a class, i.e., be of some type. How could you find out what the type of a given
type is? Does the answer depend on the given type?

## $ \S 2 $ The boolean type `bool` 

### $ 2.1 $ Description of the boolean type

The __boolean__ type, denoted by `bool` in Python, consists of only two objects:
`True` and `False`.  A __boolean expression__ is an expression that has a
boolean value, that is, which evaluates to either `True` or `False`. 

__Example:__

In [48]:
a = True
b = False
print(a, type(a))
print(b, type(b))

True <class 'bool'>
False <class 'bool'>


The boolean type supports the three basic __logical operators__ `and`, `or` and
`not` (formally called _conjunction_, _disjunction_ and _negation_,
respectively).  Note that the former two are _binary_ operators (i.e., requiring
two arguments), whereas the latter is a _unary_ operator (it works on a single
boolean argument).

📝 It can be shown that any boolean function (a function that takes a fixed, finite
number of boolean arguments and returns a boolean value) can be built solely
from these three operators. In fact, `and` and `not` suffice. Going
even further, "nand" by itself suffices, where by definition
$ x \ \texttt{nand}\ y = \texttt{not}\big(x\ \texttt{and}\ y) $, although there
is no built-in $ \texttt{nand} $ operator in Python.

__Exercise:__ By considering all pairs that can be formed from the values `True`
and `False`, construct the truth tables of the `and` and `or` operators. _Hint:_
To reduce the amount of text that needs to be typed, introduce two variables
`t` and `f` that have `True` and `False` as their values.

Booleans are extremely important in any programming language because they enable
the conditional execution of pieces of code. They usually appear in the
conditional tests of `if-elif-else` or `while` constructs to be considered later,
or as the result of comparisons. However, they can be useful in several other
situations as well.

__Exercise:__ Let $ t $ and $ f $ be variables having the values `True` and
`False`, respectively. What is the value of the following boolean expressions?

(a) `not(t)`

(b) `not t`

(c) `not(t and f)`

(d) `not t and f`

(e) `(not t) or (not f)`

(f) `not t or not f`

(g) `not not t`

(h) `t and not f`

(i) `t and f or t`


### $ 2.2 $ Comparison operators <a name="comparison"></a>

The following binary comparison operators can be applied to several types of
objects. Each of them yields either `True` or `False` as output.

| Operator   | Translation                |
| :--------  | :---------                 |
| `==`       | Equal to                   |
| `!=`       | Not equal to               |
| `<`        | Less than                  |
| `>`        | Greater than               |
| `<=`       | Less than or equal to      |
| `>=`       | Greater than or equal to   |


<div class="alert alert-warning">It is a common mistake for beginners to confuse the assignment operator <code>=</code> with the equality operator <code>==</code>. This can lead to syntactical errors (errors in the code structure, which are caught by the interpreter) or, even worse, semantic errors (discrepancies between the coder's intentions and the actual results, which are often not detected by the interpreter).</div>

📝 If a comparison is made between two values of different types, then the interpreter will first try to convert them to a common type. For example, if one compares an integer to a float, then the integer is first converted to a float.

__Exercise:__ Predict and explain the output of the following statements:

(a) `1 > 2`

(b) `1 == 0.9999999`

(c) `1 == 0.99999999999999999`

(d) `2 != 1 + 1`

(e) `2 = 1 + 1`

(f) `True != False`

(g) `3**5 >= 4**4`

(h) `"small" < "large"`

__Exercise:__ Let $ a = 8 $, $ b = 7.99 $ and $ c = -23 $. What do the following boolean expressions evaluate to?

(a) `c <= a and a >= b`

(b) `c < b < a `

(c) `c != b > a`

(d) `a <= a >= a`

In [16]:
a = 8
b = 7.99
c = -23

📝 In Python, any value of a built-in type can be interpreted as a Boolean. 
As a rule of thumb, for numerical types (`int`, `float`, `complex`), $ 0 $
is the only number that evaluates to `False`. For sequential types
(such as `str`, `list` or `tuple`), only empty objects are considered `False`.
In other words, `True` and `False` are not the only objects considered
true and false!  We can inspect how a value will be interpreted as a Boolean by
explicitly converting it to the type `bool`:

In [1]:
print(bool(0))
print(bool(1 + 3j))
print(bool(""))
print(bool("Thou wast not born for death, immortal Bird!"))

False
True
False
True


## $ \S 3 $ `None`

The type `Nonetype` consists of the single object `None`, which is used to
indicate a null value, i.e., "nothing"; it is similar to `null` in other languages,
such as C.

`None` is frequently used as a placeholder in the case where a function has no
values to return or to indicate that some object is empty or missing. For
example, a call to the `print` function always returns `None` as its output,
regardless of its argument.

* `None` is not the same as $ 0 $; it is not possible to use it
  within arithmetic expressions.
* `None` is not a string, empty or otherwise. However, if it is printed, 'None'
  is displayed.
* `None` is not the same as `False`, but it is evaluated to `False` when appearing
  in a conditional test.

__Exercise:__ Let `x = None`. Check the output of the following statements
(trying to explain some of them would require more knowledge of Python than we
have at this point):

(a) `x`

(b) `print(x)`

(c) `x != True`

(d) `x != False`

(e) `x != 0`

(f) `not x`

In [11]:
x = None

## $ \S 4 $ Numerical types and operators

Python supports three built-in data types for representing numbers:
* `int`, or __integer__ type, for integers such as $ -1 $, $ 2 $, $ 0 $ or $ 53 $;
* `float`, or __floating-point__ type, for floating-point numbers (intuitively,
  numbers with a finite decimal expansion) such as $ 3.1415 $, $ 2.0 $ or
  $ -.450 $;
* `complex`, or __complex__ type, for complex numbers such as $ 2 + 3i $ or $
  3.14 - 43.5 i$.

Each of these types also supports the equality `==` and inequality `!=`
operators, the basic arithmetic operators `+`, `-`, `*` (multiplication), `/`
(division) and a few others to be discussed below. Integers and floats (but not
complex numbers!) can also be compared in size through `<`, `>`, `<=` and `>=`.

## $ \S 5 $ The type `int` of integers

Unlike in many other languages, there is only one type for representing
integers: `int`. Moreover, these integers can in principle be of any size (a
limit is imposed only by the memory capacity of the computer).

In [3]:
a = 4
b = -3
print(a, type(a))    # Checking if a is of type int.
print(b, type(b))

4 <class 'int'>
-3 <class 'int'>


In [1]:
# Verifying if the author lied about there being no bound to
# the possible values of integers:
print("2 raised to the 64:", 2**64)
print("2 raised to the 256:", 2**256)
print("2 raised to the 1024:", 2**1024)

2 raised to the 64: 18446744073709551616
2 raised to the 256: 115792089237316195423570985008687907853269984665640564039457584007913129639936
2 raised to the 1024: 179769313486231590772930519078902473361797697894230657273430081157732675805500963132708477322407536021120113879871393357658789768814416622492847430639474124377767893424865485276302219601246094119453082952085005768838150682342462881473913110540827237163350510684586298239947245938479716304835356329624224137216


__Exercise:__ Can you predict the value of the expressions below when $ x = 3 $ and $ y = 2 $? What if $ y = - 2 $?

In [ ]:
# x = 3
# y = 2
print(x + y)      # Adding x and y.
print(x - y)      # Subtracting y from x.
print(x * y)      # Multiplying x and y.
print(x // y)     # Quotient of x by y (integer division).
print(x % y)      # Remainder of the division of x by y.
print(y**x)       # Raising y to x; may return a float.
print(x**y)       # Raising x to y; may return a float.
print(x / y)      # Division of x by y; always returns a float.

__Exercise:__ Let $ x = 64 $. What do the statements `x**0.5` and `x**(1/3)` yield? Explain.

__Exercise:__ What is $ 0^0 $ according to Python? 

## $ \S 6 $ The type `float` of floating-point numbers

Irrational numbers such as $ \pi $ or $ e $ can never be represented exactly on
a machine, regardless of whether the decimal or binary system is used,
because this would require an infinite amount of digits to be stored.
Floating-point numbers provide approximate representations of real numbers;
their type is called `float`. Using the conventional notation, a floating-point
number is characterized by the use of a decimal point `.` to separate its
integer and fractional part.

__Example:__

In [12]:
x = 3.14         # We can recognize that x and y are of type
y = -2.71        # float because of the decimal points.
z = 19.
w = .456

print(type(z))   # Verifying that z is of type float.
print(x**y)      # Raising x to y.
print(x / x)     # Dividing x by x.

<class 'float'>
0.04501120645942232
1.0


📝 Following the IEEE 754 standard, in Python floating-point numbers are
represented as $ 64 $-bit, double-precision values. The values that any
floating-point number can take on are restricted to the approximate range
between $ \pm 1.8 \cdot 10^{308} $.  If some expression yields a value that
exceeds these bounds, then an _overflow error_ or some other unexpected quantity
may result. Similarly, a nonzero number whose absolute value is smaller than $
2.3 \cdot 10^{-308} $ cannot be represented exactly; it may be rounded to $ 0.0 $,
which in turn can trigger an error (such as division by zero) depending on the
situation.


__Exercise:__ Let $ r = 3.0 $ and $ s = 1.2 $. Guess and then explain the values
that are printed below.

In [ ]:
# r = 3.0
# s = 1.2
print(r + s)      # Adding r and s.
print(r - s)      # Subtracting s from r.
print(r * s)      # Multiplying r and s.
print(r / s)      # Division of r by s.
print(r**s)       # Raising r to s.
print(r // s)     # How many times does s fit into r? Returns a float.
print(r % s)      # And what is the remainder? Returns a float.

__Exercise:__ What happens if you try to divide by $ 0 $?

A floating-point number can also be written in the __scientific__ or __exponential notation__ using `e`. In this notation, the number to the left of `e` is multiplied with $ 10 $ raised to the power to the right of `e`.

__Exercise:__ Which real numbers are represented by the following floats?

(a) `3.14159e0`

(b) `1.23e2`

(c) `2e-1`

(d) `1.e-2`

(e) `10e-1`

(f) `4.56e-2`

(g) `.789e5`

⚠️ Note that `10e1` does not represent $ 10^1 = 10 $, but rather
$$
10 \times 10^1 = 10^2 = 100\,.
$$
In other words, the `e` in the scientific notation does not stand
for the exponentiation operation, which is instead denoted by `**`.

⚠️ Scientific notation can only be used with constant values, not variables, as illustrated in the exercise below. 

__Exercise:__ Let `x = 3.14` and `y = 2`. Consider the following three ways to
get the interpreter to compute the value of $ 3.14 \times 10^2 $ and explain
their results:

(a) `print(3.14e2)`

(b) `print(xe2)`

(c) `print(3.14ey)`

In [15]:
x = 3.14
y = 2

📝 The expression "floating-point" refers to the fact that, when expressed
in scientific notation, the position of the decimal point of a number of this
type can "float" anywhere between its digits, i.e., it is not bound to the
position to the right of the digit representing units. In particular, the
same number can be represented in several ways, e.g.:
$$
3.1415 \times 10^0 = 3141.5 \times 10^{-3} = 31415 \times 10^{-4}
$$

__Exercise:__ Explain why `2023**100` evaluates to an integer while `2020.1**100` yields an overflow error. _Hint:_ See the comments in $ \S 5 $.

__Exercise:__ Evaluate `2.0 + 2.0 - 2.0` and `2.0 + 2.0e16 - 2.0e16` and try to explain why the results are different. (We will return to this problem when we study floating point arithmetic later.)

## $ \S 7 $ Type conversion

We can use `int`, `float` and `complex` as _functions_ to convert a value to
the corresponding type in certain cases (complex numbers are discussed in $ \S
10 $). Similarly, any value of a numerical type can be converted to a string by
using `str` as a function (strings will be discussed in depth in the next
notebook). More precisely:

* Any integer may be converted into a floating-point or complex number;
* Any floating-point number can be converted into a complex number;
* Any numerical value can be converted into a string;
* Conversely, a string can also be converted into a numerical type, provided
  that its literal represents a valid number of the intended type.

__Exercise:__ Suppose that $ a = 2 $. If we run the following statements in
sequence, what are the values and types of the corresponding outputs?

(a) `float(a)`

(b) `complex(a)`

(c) `str(a)`

(d) `a`

📝 Note that in any of these cases, the original object remains the same and a
_new_ object having the specified type is created. For instance, in the
preceding exercise, after all the statements have been run, $ a $ is still of
type `int`.

Applying `int` to a floating-point number $ x $ _truncates_ it, that is,
discards the fractional part of $ x $.

__Exercise:__ Compute the outputs of:

(a) `int(2)`

(b) `int(2.3)`

(c) `int(2.8)`

(d) `int(-2.3)`

(e) `int(-2.8)`

To round $ x $ to the nearest integer, we use `round(x)`. More generally,
`round(x, <number of digits>)` _rounds_ a floating-point number $ x $ to a precision
within the specified number of digits after the decimal point.

__Exercise:__ Let the variable $ b $ have the value $ 2.654321 $. What are the values of:

(a) `int(b)`

(b) `round(b)`

(c) `round(b, 3)`

(d) `round(b, 0)`

(e) `round(b, 9)`

(f) `round(-b)`

(g) `int(-b)`

In [103]:
b = 2.654321

## $ \S 8 $ Arithmetic operations
In summary, Python supports the following binary arithmetic operators:

| Operator  | Operation          | Types allowed              |
| :-------- | :---------         | :------------------------  |
| `+`       | Addition           |  `int`, `float`, `complex` |
| `-`       | Subtraction        |  `int`, `float`, `complex` |
| `*`       | Multiplication     |  `int`, `float`, `complex` |
| `/`       | Division           |  `int`, `float`, `complex` |
| `**`      | Exponentiation     |  `int`, `float`, `complex` |
| `//`      | Integer division   |  `int`, `float`            |
| `%`       | Modulo (remainder) |  `int`, `float`            |

<a name="table 1"></a>

The latter two operators are defined as follows. If $ x $ and $ y $ are
integers or floating-point numbers, then:
* ` x // y` is the largest integer $ n $ such that $ n y \le x $ (equivalently, it
is the floor of $ \frac{x}{y} $).
* `x % y` is the remainder $ x - ny $. In other words, it is defined by
the equation:
$$
\mathtt{x\ =\ y\ *\ (x\ //\ y) + x\ \%\ y}
$$

These operators are usually applied only when $ x $ and $ y $ are both positive, but
the definitions above work no matter the signs of $ x $ and $ y $.

📝 Note the similarity of the symbol `%` to the symbol $ \div \,$ for the division operator.

__Exercise:__ Predict the values of the following expressions and then check your answers:

(a) `7 * 2`

(b) `7 ** 2`

(c) `7 // 2`

(d) `7 / 2`

(e) `7 % 2`

__Exercise:__ Let $ x = 11 $, $ y = 3 $, $ s = 2.51 $ and $ t = 8.53 $. What is the output of the statements below?

(a) `x // t`

(b) `x % t`

(c) `t // y`

(d) `t % y`

(e) `t // s`

(f) `t % s`

📝 Applying `%` with second argument $ 1 $ results in the fractional part of the
(positive) first argument, as in the following example.

In [13]:
3.14159 % 1

0.14158999999999988

Besides these binary arithmetic operators, we can also make use of the following functions when working with numerical values:
* `abs`, the **absolute value** function, which takes a single numerical argument of any numerical type (`int`, `float` or `complex`) and returns its absolute value (or modulus, in the case of complex numbers);
* `max` and `min`, which can take any number of arguments of type
  `float` or `int` separated by commas and return their __maximum__ and
  __minimum__, respectively.

__Exercise:__ Determine the output of the following statements:

(a) `abs(-2.71828)`

(b) `max(-5.3, 2, 10.45, -23, 0)`

(c) `min(-5.3, 2, 10.45, -23, 0)`

Other common mathematical functions such as the exponential, the logarithm and trigonometric functions can be used after importing the `math` or `numpy` modules (more about this later).

## $ \S 9 $ Compound assignment operators

Corresponding to each of the arithmetic operators considered above, Python
provides the following __compound assignment operators__ which can be used to
perform an operation in-place on a variable provided as its first operand:

    +=, -=, *=, /=, **=, //=, %=

Each of these combines a binary operation with an assignment.
Their meaning can be gleaned from the following examples.

| Example             | Equivalent statement |
| :--------           | :---------           |
| `x += 3.14`         | ` x = x + 3.14`      |
| `x -= 2 / 3`        | `x = x - (2 / 3)`    |
| `x *= 2`            | `x = x * 2`          |
| `x /= 0.5`          | `x = x / 0.5`        |
| `x **= 1 / 2`       | `x = x**(1 / 2)`     |
| `x //= z`           | `x = x // z`         |
| `x %= y**2`         | `x = x % (y**2)`     |

In each case, the _current_ value of the variable $ x $ is used to compute
the expression on the right, which becomes the new value of $ x $. Note also
that in order to interpret compound assignments correctly, the expression on the
right side must always be parenthesized. Thus, `x *= y + 1` is equivalent to `x = x *
(y + 1)`, not `x = x * y + 1`.

__Exercise:__ Let $ x = 7 $ and $ y = 5 $. Determine the
values of $ x $ and $ y $ after each of the following statements is run through
the interpreter in sequence:

(a) `x -= y + 1`

(b) `y += x`

(c) `y **= x`

(d) `y //= x`

(e) `y %= x`

(f) `x **= 1/2`

In [ ]:
x = 7
y = 5

📝 Actually `a = a + b` and `a += b` are not exactly equivalent. The latter
modifies the value of $ a $ _in place_, hence is generally more efficient; the
former creates a _new object_ with the value `a + b` and then assigns it to $ a $.
However, for all of our purposes this slight difference can be ignored.

## ⚡ $ \S 10 $ The type `complex` of complex numbers

We won't need to use complex numbers in this course, hence this section is
optional. A number of type `complex` can be represented in Python as a pair of
floats (its real and imaginary parts) using the special notation indicated in
the following examples.

⚠️ In Python the imaginary unit is denoted by $ j $ instead of the more usual
$ i $. 

In [ ]:
w = 1.23 + 4.56j    # Note the syntax: we write '4.56j', not '4.56 * j'.
print(w, type(w))   # Checking the type of w.

z = 4 + 1j       
print(z, type(z))

print(z + w)

(1.23+4.56j) <class 'complex'>
(4+1j) <class 'complex'>
(5.23+5.56j)


⚠️ By itself, `j` gets interpreted as a variable whose name is $ j $.  In order
to indicate that we want a complex number, `j` must be immediately preceded by a
number.

In [ ]:
print(j)    # Will result in an error if there is no variable named 'j'.

NameError: name 'j' is not defined

In [ ]:
# Even if the imaginary part is zero, it must still be included explicitly to
# indicate that we are dealing with a complex number instead of a float
# or an int:
a = 1 + 0j
print(a, type(a))
# However, in case the real part is 0, we may omit it:
b = 1j
print(b, type(b))

(1+0j) <class 'complex'>
1j <class 'complex'>


📝 Each complex number has special attributes `real` and `imag` to
represent its real and imaginary parts respectively, and a special method
`conjugate` to compute its conjugate. 

In [ ]:
# Complex number with real part 0 and imaginary part 1.0:
w = 1.0j
print(w)    
print(w.real, type(w.real))
print(w.imag, type(w.imag))

1j
0.0 <class 'float'>
1.0 <class 'float'>


In [ ]:
# Complex number with real part 0 and imaginary part 1:
z = 0 + 1j
# Printing z and the type of its real and imaginary parts:
print(z)    
print(z.real, type(z.real))
print(z.imag, type(z.imag))
# Although we used the integers 0 and 1 as the real and imaginary parts, the
# resulting complex number still has real and imaginary parts of type float!

1j
0.0 <class 'float'>
1.0 <class 'float'>


In [ ]:
print(z.conjugate(), type(z.conjugate()))

-1j <class 'complex'>


__Exercise:__ Suppose that the statements `z = 1 + 1j` and `w = 1 -1j` have just been interpreted.
What is the value and type of the following expressions?

(a) `z + w`

(b) `z - w`

(c) `z**2`

(d) `z**4`

(e) `z * w`

(f) `z / w`

(g) `w != z`

(h) `z + w <= 2 * (z + w)`

__Exercise:__ Use Python to guess the value of $ e^{\pi i} $. _Hint:_ Use approximations for $ e $ and $ \pi $.

🚫 Trying to compare two complex numbers or to convert a complex number into an
integer or floating-point number will result in a `TypeError`.

__Exercise:__ Suppose that the statement `z = 2.0 + 0j` has just been run through the interpreter. Describe the outputs of:

(a) `int(z)`

(b) `float(z)`

(c) `str(z)`

(d) `complex('2.0 + 0j')`

(e) `complex('2.0')`